javacutil/src/main/java/org/checkerframework/javacutil/TreeUtils.java

```diff
@@ -582,28 +579,43 @@ public static ExecutableElement constructor(NewClassTree tree) {
         }
 
         JCNewClass newClassTree = (JCNewClass) tree;
+        Element e = newClassTree.constructor;
+        return (ExecutableElement) e;
+    }
 
-        if (tree.getClassBody() != null) {
-            // anonymous constructor bodies should contain exactly one statement
-            // in the form:
-            //    super(arg1, ...)
-            // or
-            //    o.super(arg1, ...)
-            //
-            // which is a method invocation (!) to the actual constructor
-
-            // the method call is guaranteed to return nonnull
-            JCMethodDecl anonConstructor =
-                    (JCMethodDecl) TreeInfo.declarationFor(newClassTree.constructor, newClassTree);
-            assert anonConstructor != null;
-            assert anonConstructor.body.stats.size() == 1;
-            JCExpressionStatement stmt = (JCExpressionStatement) anonConstructor.body.stats.head;
-            JCTree.JCMethodInvocation superInvok = (JCMethodInvocation) stmt.expr;
-            return (ExecutableElement) TreeInfo.symbol(superInvok.meth);
-        } else {
-            Element e = newClassTree.constructor;
-            return (ExecutableElement) e;
+    /**
+     * Given an anonymous class instantiation via {@code new}, returns symbol for the super
+     * constructor, i.e. the one that gets invoked in the anonymous constructor (JLS 15.9.5.1)
+     *
+     * @param tree the anonymous constructor invocation
+     * @return the {@link ExecutableElement} corresponding to the super constructor called in {@code
+     *     tree}
+     */
+    public static ExecutableElement anonymousSuperConstructor(NewClassTree tree) {
+        assert tree.getClassBody() != null;
+
+        if (!(tree instanceof JCTree.JCNewClass)) {
```

Comment:
I would remove this check. The cast below would fail and give an error - and so many other things would have failed first.